common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,208 @@
+     *
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R>    The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    public static <V, R> Future<R> flatMap(Future<V> future, Function<V, Future<R>> mapper) {
+        Objects.requireNonNull(future, "The future cannot be null.");
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
+        Promise<R> promise = future.executor().newPromise();
+        future.addListener(new FlatMapper<>(promise, mapper));
+        if (!future.isSuccess()) {
+            // Propagate cancellation if future is either incomplete or failed.
+            // Failed means it could be cancelled, so that needs to be propagated.
+            promise.addListener(future, PromiseNotifier::propagateCancel);
+        }
+        return promise;
+    }
+
+    private Futures() {
+    }
+
+    @SuppressWarnings("unchecked")
+    static <R> FutureContextListener<Promise<R>, Object> passThrough() {
+        return (FutureContextListener<Promise<R>, Object>) (FutureContextListener<?, ?>) PASS_THROUGH_CONSTANT;
+    }
+
+    static <A, B> void propagateUncommonCompletion(Future<? extends A> completed, Promise<B> recipient) {
+        if (completed.isCancelled()) {
+            // Don't check or log if cancellation propagation fails.
+            // Propagation goes both ways, which means at least one future will already be cancelled here.
+            recipient.cancel(false);
+        } else {
+            Throwable cause = completed.cause();
+            tryFailure(recipient, cause, LOGGER);
+        }
+    }
+
+    private static class PassThrough<R> implements FutureContextListener<Promise<R>, Object> {
+        @Override
+        public void operationComplete(Promise<R> recipient, Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    R result = (R) completed.getNow();
+                    recipient.trySuccess(result);
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class CallableMapper<R, T> implements Callable<R> {
+        private final Future<T> future;
+        private final Function<T, R> mapper;
+
+        CallableMapper(Future<T> future, Function<T, R> mapper) {
+            this.future = future;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public R call() throws Exception {
+            return mapper.apply(future.getNow());
```

Comment:
Add an `assert future.isDone()` ?

Reply:
I think that's not necessary since this is in the same "component" where such checks are done we get to this line.